### DISCUSSION TITLE
Developing and not losing pro features

### DISCUSSION THREAD

**Discussion by wilbur4321** [tiny-pilot/tinypilot#781] [General]
As a customer of the Pro version, what do I need to do to work on the code to add new features -- I'm wanting to add client-side macros to the webui -- without losing the Pro features?

**Comment by mtlynch** [comment]
Thanks for reaching out, @wilbur4321. That would be a great feature.

You should check out our [contributor guidelines](https://github.com/tiny-pilot/tinypilot/blob/master/CONTRIBUTING.md) to understand the contribution process in general.

Since this is a nontrivial feature, the next step is to create an issue to sketch our your proposal. In this case, we have an existing bug (#481), so you'd sketch out details there of what the UI would be for this feature and how we'd store the settings.

In terms of development, you have two options:

1. Develop on your instance of TinyPilot Pro and backport to this repo to make PRs
1. Develop on an dev instance with TinyPilot Community installed

I'd personally recommend (2). This feature is one that's pretty easy to develop [on a dev machine](https://github.com/tiny-pilot/tinypilot/blob/master/CONTRIBUTING.md#setting-up-a-development-environment) rather than on real hardware. Once you have it working on your dev machine, you can create a patch and apply it to your Pro instance. You'll may have to resolve a few merge conflicts, but this feature should be fairly independent from the rest of the codebase, so there shouldn't bee too much work in porting the changes forward.

If you go route (1), make sure that you save your code to a remote repo, especially before applying a TinyPilot update. TinyPilot updates overwrite any local code changes you have on the system.
